File: samples/bot-teams-authentication/nodejs/bots/dialogBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { TeamsActivityHandler } = require('botbuilder');

/**
 * DialogBot class that extends TeamsActivityHandler to handle Teams activities.
 */
class DialogBot extends TeamsActivityHandler {
    /**
     * Creates an instance of DialogBot.
     * @param {ConversationState} conversationState - The state management object for conversation state.
     * @param {UserState} userState - The state management object for user state.
     * @param {Dialog} dialog - The dialog to be run by the bot.
     */
    constructor(conversationState, userState, dialog) {
        super();
        if (!conversationState) throw new Error('[DialogBot]: Missing parameter. conversationState is required');
        if (!userState) throw new Error('[DialogBot]: Missing parameter. userState is required');
        if (!dialog) throw new Error('[DialogBot]: Missing parameter. dialog is required');

        this.conversationState = conversationState;
        this.userState = userState;
        this.dialog = dialog;
        this.dialogState = this.conversationState.createProperty('DialogState');

        this.onMessage(this.handleMessage.bind(this));
    }

    /**
     * Handles incoming message activities.
     * @param {TurnContext} context - The context object for the turn.
     * @param {function} next - The next middleware function in the pipeline.
     */
    async handleMessage(context, next) {
        console.log('Running dialog with Message Activity.');

        // Run the Dialog with the new message Activity.
        await this.dialog.run(context, this.dialogState);

        await next();
    }

    /**
     * Override the ActivityHandler.run() method to save state changes after the bot logic completes.
     * @param {TurnContext} context - The context object for the turn.
     */
    async run(context) {
        await super.run(context);

        // Save any state changes. The load happened during the execution of the Dialog.
        await this.conversationState.saveChanges(context, false);
        await this.userState.saveChanges(context, false);
    }
}

module.exports.DialogBot = DialogBot;

File: samples/bot-teams-authentication/python/bots/dialog_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import ConversationState, UserState, TurnContext
from botbuilder.core.teams import TeamsActivityHandler
from botbuilder.schema import InvokeResponse
from helpers.dialog_helper import DialogHelper
from botbuilder.dialogs import Dialog
import logging
import traceback

class DialogBot(TeamsActivityHandler):
    def __init__(
        self,
        conversation_state: ConversationState,
        user_state: UserState,
        dialog: Dialog,
    ):
        # Initializes the DialogBot with conversation state, user state, and main dialog.
        if conversation_state is None:
            raise Exception(
                "[DialogBot]: Missing parameter. conversation_state is required"
            )
        if user_state is None:
            raise Exception("[DialogBot]: Missing parameter. user_state is required")
        if dialog is None:
            raise Exception("[DialogBot]: Missing parameter. dialog is required")

        self.conversation_state = conversation_state
        self.user_state = user_state
        self.dialog = dialog

    async def on_turn(self, turn_context: TurnContext):
        # Handles every turn of the bot and saves any state changes.
        try:
            await super().on_turn(turn_context)

            # Save any state changes that might have occurred during the turn.
            await self.conversation_state.save_changes(turn_context, False)
            await self.user_state.save_changes(turn_context, False)

        except Exception as e:
            logging.error(f"Error in on_turn: {e}")
            traceback.print_exc()
            await turn_context.send_activity("Sorry, something went wrong processing your message.")
            
    async def on_message_activity(self, turn_context: TurnContext):
        # Handles message activities and manages dialog flow including 'clear' command.
        logging.info("on_message_activity triggered.")
        try:
            text = (turn_context.activity.text or "").strip().lower()
            dialog_state_accessor = self.conversation_state.create_property("DialogState")

            if text == "clear":
                logging.info("'clear' command received. Attempting to reset dialog state.")
                dialog_state = await dialog_state_accessor.get(turn_context)

                if dialog_state:
                    await dialog_state_accessor.delete(turn_context)
                    logging.info("Dialog state deleted.")
                    await turn_context.send_activity("Dialog state cleared. You can start again.")
                else:
                    logging.info("No dialog state found to delete.")
                    await turn_context.send_activity("No active dialog state found to clear.")
                return

            await DialogHelper.run_dialog(
                self.dialog,
                turn_context,
                dialog_state_accessor,
            )

        except Exception as e:
            logging.error(f"Error in on_message_activity: {e}")
            traceback.print_exc()
            await turn_context.send_activity("An error occurred handling your message.")

    async def on_invoke_activity(self, turn_context: TurnContext):
        # Handles invoke activities such as token exchange or state verification.
        try:
            logging.info(f"Incoming activity type: {turn_context.activity.type}")
            logging.info(f"Activity name: {turn_context.activity.name}")
    
            if turn_context.activity.name == "signin/tokenExchange":
                logging.info("Handling signin/tokenExchange")
    
                # Token exchange success, continue with dialog
                await DialogHelper.run_dialog(
                    self.dialog,
                    turn_context,
                    self.conversation_state.create_property("DialogState"),
                )
                return InvokeResponse(status=200)
    
            elif turn_context.activity.name == "signin/verifyState":
                logging.info("Handling signin/verifyState")
    
                await DialogHelper.run_dialog(
                    self.dialog,
                    turn_context,
                    self.conversation_state.create_property("DialogState"),
                )
    
                return InvokeResponse(status=200)
    
            return await super().on_invoke_activity(turn_context)
    
        except Exception as e:
            logging.error(f"Error in on_invoke_activity: {e}")
            traceback.print_exc()
            return InvokeResponse(status=500)
